Add Test Monitor file attachment example

BKnight760 opened the pull request:
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/<reponame>/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?
This adds a Test Monitor C# client example to show how to attach files to test results.

### Why should this Pull Request be merged?
Attaching files is a specific use-case that customers want to know how to do.

### What testing has been done?
The example has been executed and verified it correctly publishes a test result and attaches a single file.

BKnight760 committed:
Add file attachment example.

BKnight760 committed:
Add FileAttachment README.md and associated image.

BKnight760 committed:
Update the parent project README.md files.

BKnight760 committed:
Update file comment.

jjaco2 commented on examples/testmonitor/README.md:
```diff
@@ -28,4 +28,5 @@ Test Monitor Examples
 ------------
 
 - [Results](results): Demonstrates how to use the SystemLink Test Monitor API to create test results and steps.
+- [File Attachment](fileattachment): Demonstrates how to use the SystemLink Test Monitor API to attach files that are uploaded to the SystemLink server using the SystemLink File API.
```
Attach them to a test result, right? Maybe

> to attach files from the SystemLink server to a test result using the SystemLink File API.

Also curious how attached we are to specifying "SystemLink" before API names. I realize it might be more necessary in a non-NI-specific space.

jjaco2 replied:
Attach them to a test result, right? Maybe 'to attach files from the SystemLink server to a test result using the SystemLink File API.'

Also curious how attached we are to specifying 'SystemLink' before API names. I realize it might be more necessary in a non-NI-specific space. Applies throughout.

jjaco2 commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,82 @@
+{
+    /// <summary>
+    /// Example for the SystemLink Test Monitor API that creates a test result,
+    /// uploads a file to the SystemLink server and attaches the file to the
```
add comma (NI style and consistency)

jjaco2 commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,82 @@
+{
+    /// <summary>
+    /// Example for the SystemLink Test Monitor API that creates a test result,
+    /// uploads a file to the SystemLink server and attaches the file to the
```
> SystemLink server,

Add comma for NI style.

jjaco2 replied:
didn't mean to make that look like a quote. it's my suggested revision.

jjaco2 commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,82 @@
+                "stream.txt",
+                Encoding.UTF8.GetBytes("ABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+
+            // Add the file Id to the test result's list of attached files.
```
> file Id 

ID since we're not talking about the label itself

jjaco2 commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,82 @@
+            // Add the file Id to the test result's list of attached files.
+            resultData.FileIds.Add(fileId);
+
+            // Set the test result's status to done.
```
'test result' should work fine here.

jjaco2 commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,82 @@
+
+        public static string UploadFileUsingStream(Core.IHttpConfiguration configuration, string fileName, byte[] fileContents)
+        {
+            // Use the FileUploader from the SystemLink File client
```
Is this different than just uploading a file? And does this exist anywhere else that's user-facing, or do we refer to it exclusively as FileUploader?

Maybe 'Upload a file using [File Uploader/FileUploader/the file uploader]'

BKnight760 replied:
The FileUploader is just the name of a class in the client's source code.  This particular line isn't doing the upload, that's a few lines down.  The comment is just supposed to highlight that we're using a different SystemLink client API within this example.

BKnight760 replied:
How about 'Upload a file using the SystemLink File client'?

BKnight760 replied:
Yes, 'SystemLink' prefixed before every API gets a bit verbose.  When viewed as an entire project of examples, 'SystemLink' can probably be inferred from the context.  I've left them in for now, but will make the change if we want to do that throughout all of these examples for consistency.  I'd do that in another PR and scrub the entire repo.

jjaco2 commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,82 @@
+                // Reset the stream position to the beginning to upload from the beginning.
+                memoryStream.Position = 0;
+
+                // Upload the file to the SystemLink server and get the uploaded file's Id.
```
> the uploaded file's Id.

the ID for the uploaded file.

jjaco2 commented on examples/testmonitor/fileattachment/README.md:
```diff
@@ -0,0 +1,32 @@
+=================
+
+This is an example console application demonstrating how to use the
+SystemLink Test Monitor API attach a file to a test result.
```
API to attach

jjaco2 commented on examples/testmonitor/fileattachment/README.md:
```diff
@@ -0,0 +1,32 @@
+About the Example
+-----------------
+
+This example creates a single test result and attaches a file.  It uses the SystemLink File client to upload an in-memory byte stream as the contents of a file.  The file client allows different methods of uploading files, but for the purpose of this example the stream upload is utilized.  The resulting file Id is added to the test result data and updated on the server.
```
Throughout: single space after periods

BKnight760 replied:
At least I'm consistent, right?  lol.  I'll single-space them all.  Force of habit.

jjaco2 commented on examples/testmonitor/fileattachment/README.md:
```diff
@@ -0,0 +1,32 @@
+About the Example
+-----------------
+
+This example creates a single test result and attaches a file.  It uses the SystemLink File client to upload an in-memory byte stream as the contents of a file.  The file client allows different methods of uploading files, but for the purpose of this example the stream upload is utilized.  The resulting file Id is added to the test result data and updated on the server.
```
> but for the purpose of this example the stream upload is utilized

this example uses the stream upload method.

jjaco2 replied:
> The resulting file Id

The resulting file ID

jjaco2 commented on examples/testmonitor/fileattachment/README.md:
```diff
@@ -0,0 +1,32 @@
+
+### Sample output
+
+The data published to the SystemLink server can be viewed in the [Test Monitor Web Application](https://localhost/#testmonitor).  The test result details and the attached files can be explored using a web browser:
```
View the data published to the SystemLink server in the [Test Monitor Web Application] (https://localhost/#testmonitor). Explore test result details and attached files using a web browser:

BKnight760 committed:
Update documentation per review feedback.

BKnight760 committed:
Documentation change per review feedback.

alexweav commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,91 @@
+            var testDataManager = new TestDataManager(configuration);
+
+            // Intialize the random number generator.
+            var random = new Random();
```
Is this rng used anywhere?

spanglerco replied:
It's probably random whether or not it gets used.

alexweav commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <param name="fileName">The name for the uploaded file.</param>
+        /// <param name="fileContents">An array of bytes to send for the contents of the file.</param>
+        /// <returns>The ID of the file that was uploaded.</returns>
+        public static string UploadFileUsingStream(Core.IHttpConfiguration configuration, string fileName, byte[] fileContents)
```
```suggestion
        public static string UploadFileData(IHttpConfiguration configuration, string fileName, byte[] fileContents)
```
Rename is a nitpick, ignore if you disagree. Also add `using NationalInstruments.SystemLink.Clients.Core;` up top so the removal of `Core.` doesn't cause any errors.

spanglerco commented on examples/README.md:
```diff
@@ -61,4 +61,5 @@ API Examples
-- [Results](testmonitor/results): Demonstrates how to use the SystemLink Test Monitor API to publish test results to the server.
\ No newline at end of file
+- [Results](testmonitor/results): Demonstrates how to use the SystemLink Test Monitor API to publish test results to the server.
+- [File Attachment](testmonitor/fileattachment): Demonstrates how to use the SystemLink Test Monitor API to attach files that are uploaded to the SystemLink server using the SystemLink File API.
```
It's also fine to use hyphens if you think `file-attachment` would be better than `fileattachment`. I don't feel strongly either way.

BKnight760 replied:
Renamed to file-attachment.

spanglerco commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,91 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using NationalInstruments.SystemLink.Clients.TestMonitor;
+using NationalInstruments.SystemLink.Clients.File;
```
```suggestion
using NationalInstruments.SystemLink.Clients.File;
using NationalInstruments.SystemLink.Clients.TestMonitor;
```

spanglerco commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,91 @@
+            var fileUploader = new FileUploader(configuration);
+
+            // Use in-memory data for the file upload.
+            using (var memoryStream = new MemoryStream())
```
Doesn't the MemoryStream constructor allow you to pass in a byte array to wrap?

BKnight760 replied:
Yes, cleaned up that method quite a bit.

spanglerco commented on examples/testmonitor/fileattachment/FileAttachment.cs:
```diff
@@ -0,0 +1,91 @@
+             * See the configuration example for how a typical application
+             * might obtain a configuration.
+             */
+            var configuration = ExampleConfiguration.Obtain(args);
```
Probably true for the previous Test Monitor example as well, but I just realized we should be doing this:
```suggestion
            var configuration = ExampleConfiguration.Obtain(args, allowCloud: false);
```

BKnight760 committed:
Rename UploadFile method.  Add using to remove need for 'Core' prefix.

Co-Authored-By: Alexander Weaver <weaver.alex.d@gmail.com>

BKnight760 committed:
Reorder usings

Co-Authored-By: Paul Spangler <spanglerco@users.noreply.github.com>

BKnight760 committed:
Set allowCloud: false on the configuration

Co-Authored-By: Paul Spangler <spanglerco@users.noreply.github.com>

BKnight760 committed:
Remove the random number generator.

BKnight760 committed:
Merge branch 'users/mvaterla/add-file-attachment-example' of https://github.com/ni/systemlink-client-docs into users/mvaterla/add-file-attachment-example

BKnight760 committed:
Add using statement for Clients.Core.
Rename UploadFileData method.
Use MemoryStream constructor to pass the file contents.

BKnight760 committed:
Rename directory to file-attachment.
